enhancements/ingress/haproxy-max-connections-tuning.md

```diff
@@ -0,0 +1,350 @@
+
+The administrator can patch their existing ingress controllers to
+increase the number of simultaneous connections. 
+
+For example, patching the `default` ingresscontroller to support
+150000 simultaneous connections.
+
+```sh
+$ oc patch ingresscontroller/default --type=merge --patch '{"spec":{"tuningOptions":{"maxConnections":150000}}}' -n openshift-ingress-operator
+```
+
+Ingress controller pods will roll out with `maxconn` set to 150000.
+
+#### Story 2
+
+> As a cluster administrator, I have a node with large amounts of
+> resources (e.g., 128 cores and 1TB RAM) that I would like to handle
+> as much of my ingress as possible.
+
+To do this, the administrator can configure
+`spec.nodePlacement.nodeSelector` with labels that match the intended
+node, as well as configuring `spec.tuningOptions.maxConnections`.
+
+Specifying `spec.tuningOptions.maxConnections: -1` instructs HAProxy
+to dynamically compute the largest possible value based on the ulimits
+within the container when HAProxy starts. The nature of HAProxy's
+dynamic computation also takes into consideration what is configured
+in `haproxy.config` at that time.
+
+Additionally, if these are dedicated infrastructure nodes the `ulimit
+-n` value (i.e., maximum number of open files) can be increased by
+applying a custom tuned profile for those dedicated infrastructure
+nodes.
+
+Example: letting HAProxy compute the maximum value:
+
+```sh
+$ oc patch ingresscontroller/default --type=merge --patch '{"spec":{"tuningOptions":{"maxConnections":-1}}}' -n openshift-ingress-operator
+```
+
+Ingress controller pods will roll out with a new `maxconn` value as
+computed by HAProxy.
+
+#### Story 3
+
+> As a cluster administrator, I want to restore the default value for
+> max connections.
+
+```sh
+$ oc patch ingresscontroller/default --type=merge --patch '{"spec":{"tuningOptions":{"maxConnections":null}}}' -n openshift-ingress-operator
+```
+
+Ingress controller pods will roll out with the default setting.
+
+#### Story 4
+
+> As a cluster administrator, I would like to know what the value of
+> `maxconn` is when setting `spec.tuningOptions.maxConnections: -1`.
+
+This can be done in two stages:
+
+First set `spec.tuningOptions.maxConnections` to `-1` and let the
+router deployment roll out the new pods.
+
+```sh
+$ oc patch ingresscontroller/default --type=merge --patch '{"spec":{"tuningOptions":{"maxConnections":-1}}}' -n openshift-ingress-operator
+```
+
+We can now extract the computed value from HAProxy's built-in stats
+socket:
+
+```sh
+$ oc rsh -n openshift-ingress <ROUTER-POD> bash -c 'echo "show info" | socat /var/lib/haproxy/run/haproxy.sock stdio' | grep Maxconn
+Maxconn: 524260
+```
+
+### Risks and Mitigations
+
+#### Increased Resource Usage
+
+HAProxy builds its data structures ahead of time. If you specify a
+large value for `spec.tuningOptions.maxConnections` then that memory
+is allocated up-front when the process starts. It is never released.
```

Comment:
I was curious and I confirmed this as well. 500 backend leastconn config with 20,000 maxconn was about ~280mB. Increased to 500,000 maxconn and it spiked to 366mB.